src/riak_kv_wm_queue.erl

```diff
@@ -77,25 +93,100 @@ service_available(RD, Ctx=#ctx{riak=RiakProps}) ->
              Ctx}
     end.
 
+
+allowed_methods(RD, Ctx) ->
+    {['GET', 'POST'], RD, Ctx}.
+
 -spec malformed_request(#wm_reqdata{}, context()) ->
     {boolean(), #wm_reqdata{}, context()}.
 %% @doc Determine whether request is well-formed
-malformed_request(RD, Ctx) ->
-    QueueNameRaw = wrq:path_info(queuename, RD),
-    QueueName =
-        list_to_atom(riak_kv_wm_utils:maybe_decode_uri(RD, QueueNameRaw)),
+malformed_request(RD, Ctx) when Ctx#ctx.method =:= 'GET' ->
+    QueueName = malformed_queuename(RD),
     ObjectFormatRaw = wrq:get_qs_value(?Q_OBJECT_FORMAT, "internal", RD),
     ObjectFormat = list_to_atom(ObjectFormatRaw),
-    {false, RD, Ctx#ctx{queuename = QueueName, object_format = ObjectFormat}}.
+    {false, RD, Ctx#ctx{queuename = QueueName, object_format = ObjectFormat}};
+malformed_request(RD, Ctx) when Ctx#ctx.method =:= 'POST' ->
+    QueueName = malformed_queuename(RD),
+    KeyClockList =
+        lists:map(fun({B, K, C}) -> {B, K, C, to_fetch} end, 
+                    malformed_keyclocklist(wrq:req_body(RD))),
+    {false, RD, Ctx#ctx{queuename = QueueName, keyclocklist = KeyClockList}}.
+
+-spec malformed_queuename(#wm_reqdata{}) -> atom().
+malformed_queuename(RD) ->
+    QueueNameRaw = wrq:path_info(queuename, RD),
+    list_to_atom(riak_kv_wm_utils:maybe_decode_uri(RD, QueueNameRaw)).
+
+-spec malformed_keyclocklist(iolist()) -> list(riak_kv_replrtq_src:repl_entry()).
+malformed_keyclocklist(ReqBody) ->
+    {struct, [{<<"keys-clocks">>, KCL}]} = mochijson2:decode(ReqBody),
+    lists:reverse(lists:foldl(fun decode_bucketkeyclock/2, [], KCL)).
+
+decode_bucketkeyclock({struct, BKC}, Acc) ->
```

Comment:
Suggestions, just map over the list and raise an error if <<"key">>, <<"clock">> or other expected fields are missing.

Reply:
This has now changed so that we check for nay elements being filtered out, due to incorrect encoding, an the whole request will be malformed unless all elements are correct